x/bank/keeper.go

```diff
@@ -175,6 +175,19 @@ func addCoins(ctx sdk.Context, am auth.AccountMapper, addr sdk.AccAddress, amt s
 // SendCoins moves coins from one account to another
 // NOTE: Make sure to revert state changes from tx on error
 func sendCoins(ctx sdk.Context, am auth.AccountMapper, fromAddr sdk.AccAddress, toAddr sdk.AccAddress, amt sdk.Coins) (sdk.Tags, sdk.Error) {
+	// check if sender is vesting account
+	blockTime := ctx.BlockHeader().Time
+	vacc, ok := am.GetAccount(ctx, fromAddr).(auth.VestingAccount)
```

Comment:
Is there a way we can implement this which just implements `BaseAccount` differently and returns `SendableCoins()` for `GetCoins()`?

Reply:
Hmm... wouldn't that entail changing the `Account` interface?

Reply:
Don't believe we can without a lot of changes to bank and other modules that depend on it.

For example, bank keeper's subtractCoins method needs `GetCoins` to return the full amount in the account. Otherwise the staking handler will only allow you to delegate unlocked coins

Reply:
I actually think this is important to do though. A large premise of the SDK is that you should be able to add your own account types that can support all the features you want within your other modules. Here this is a sign that the API we provide is insufficient, or we haven't tried to make vesting fit into the correct API. We want anyone to be able to achieve the functionality they want with their custom accounts without modding baseapp like this.

Perhaps we add a "UseableCoins"  function to account that takes in the context. This would have utility outside of just vesting, e.g. for covenants and the like.